readhomer_atlas/library/models.py

```diff
@@ -1,92 +1,31 @@
 from django_extensions.db.fields.json import JSONField
+from treebeard.mp_tree import MP_Node
 
 
-class Version(models.Model):
-    """
-    urn:cts:greekLit:tlg0012.tlg001.perseus-grc2
-    """
+class Node(MP_Node):
+    # @@@ used to pivot siblings; may be possible if we hook into path field
+    idx = models.IntegerField(help_text="0-based index", blank=True, null=True)
+    # @@@ if we expose kind, can access some GraphQL enumerations
+    kind = models.CharField(max_length=255)
+    urn = models.CharField(max_length=255, unique=True)
+    ref = models.CharField(max_length=255, blank=True, null=True)
+    rank = models.IntegerField(blank=True, null=True)
+    text_content = models.TextField(blank=True, null=True)
+    metadata = JSONField(default=dict, blank=True, null=True)
 
-    urn = models.CharField(max_length=255)
-    name = models.CharField(blank=True, null=True, max_length=255)
-    metadata = JSONField(default=dict, blank=True)
-    """
-    {
-        "work_urn": "urn:cts:greekLit:tlg0012.tlg001",
-        "work_title": "Iliad",
-        "type": "edition"
-    }
-    """
-
-    class Meta:
-        ordering = ["urn"]
+    alphabet = settings.NODE_ALPHABET
 
     def __str__(self):
-        return self.name
-
-
-class Book(models.Model):
-    """
-    urn:cts:greekLit:tlg0012.tlg001.perseus-grc2:1
-    """
-
-    urn = models.CharField(max_length=255)
-    ref = models.CharField(max_length=255)
-
-    position = models.IntegerField()
-    idx = models.IntegerField(help_text="0-based index")
-
-    version = models.ForeignKey(
-        "library.Version", related_name="books", on_delete=models.CASCADE
-    )
-
-    class Meta:
-        ordering = ["idx"]
+        return f"{self.kind}: {self.urn}"
 
     @property
-    def label(self):
-        return self.label
-
-    def __str__(self):
-        return f"{self.version} [book={self.position}]"
-
-    @classmethod
-    def generate_urn(cls, version_urn, ref):
-        return f"{version_urn}{ref}"
-
-
-class Line(models.Model):
-    """
-    urn:cts:greekLit:tlg0012.tlg001.perseus-grc2:1.1
-    """
-
-    text_content = models.TextField()
-    urn = models.CharField(max_length=255)
-    ref = models.CharField(max_length=255)
-
-    position = models.IntegerField()
-    book_position = models.IntegerField()
-    idx = models.IntegerField(help_text="0-based index")
-
-    book = models.ForeignKey(
-        "library.Book", related_name="lines", on_delete=models.CASCADE
-    )
-    version = models.ForeignKey(
-        "library.Version", related_name="lines", on_delete=models.CASCADE
-    )
-
-    class Meta:
-        ordering = ["idx"]
+    def name(self):
+        return self.metadata.get("work_title")
 
     @property
-    def label(self):
-        return self.ref
-
-    def __str__(self):
-        return f"{self.version} [line_num={self.label}]"
-
-    @classmethod
-    def generate_urn(cls, version_urn, ref):
-        return f"{version_urn}{ref}"
+    def position(self):
```

Comment:
I think we can just drop this, since we're no longer using `position` to resolve lines in a GraphQL query; if we need it again in the future, I think we're going to want to denorm from the database or munge treebeard's `path` because this would likely perform too poorly in a loop to be useful in a GraphQL response 😉

Reply:
Dropped.